flink-runtime/src/test/java/org/apache/flink/runtime/source/coordinator/SourceCoordinatorAlignmentBenchmark.java

```diff
@@ -0,0 +1,97 @@
+import org.apache.flink.api.connector.source.Boundedness;
+import org.apache.flink.api.connector.source.mocks.MockSource;
+import org.apache.flink.api.connector.source.mocks.MockSourceSplit;
+import org.apache.flink.runtime.jobgraph.OperatorID;
+import org.apache.flink.runtime.operators.coordination.MockOperatorCoordinatorContext;
+import org.apache.flink.runtime.source.event.ReportedWatermarkEvent;
+
+import javax.annotation.Nonnull;
+
+import java.util.Random;
+
+import static org.apache.flink.runtime.source.coordinator.CoordinatorTestUtils.waitForCoordinatorToProcessActions;
+
+/** The benchmark of watermark alignment. */
+public class SourceCoordinatorAlignmentBenchmark {
+
+    /** The TimestampType determines how the timestamp or watermark is generated. */
+    public enum TimestampType {
+        MONOTONE_INCREASING,
+        RANDOMIZE_MILLISECONDS
+    }
+
+    private static final OperatorID OPERATOR_ID = new OperatorID(1234L, 5678L);
+    private static final Random RANDOM = new Random();
+
+    private SourceCoordinator<?, ?> sourceCoordinator;
+    private int numSubtasks;
+    private TimestampType timestampType;
+
+    public SourceCoordinatorAlignmentBenchmark() {}
+
+    public void setup(int numSubtasks, @Nonnull TimestampType timestampType) throws Exception {
+        SourceCoordinatorProvider<MockSourceSplit> provider =
+                new SourceCoordinatorProvider<>(
+                        "SourceCoordinatorProviderTest",
+                        OPERATOR_ID,
+                        new MockSource(Boundedness.BOUNDED, 2),
+                        1,
+                        new WatermarkAlignmentParams(1000L, "group1", Long.MAX_VALUE),
+                        null);
+        this.sourceCoordinator =
+                (SourceCoordinator<?, ?>)
+                        provider.getCoordinator(
+                                new MockOperatorCoordinatorContext(OPERATOR_ID, numSubtasks));
+        this.sourceCoordinator.start();
+        this.numSubtasks = numSubtasks;
+        this.timestampType = timestampType;
+
+        // Initialize the watermark for all subtasks.
+        sendReportedWatermarkToAllSubtasks();
+    }
+
+    public void teardown() throws Exception {
+        sourceCoordinator.close();
+    }
+
+    public void sendReportedWatermarkToAllSubtasks() {
+        for (int subtaskIndex = 0; subtaskIndex < numSubtasks; subtaskIndex++) {
+            sourceCoordinator.handleEventFromOperator(
+                    subtaskIndex, 0, new ReportedWatermarkEvent(generateTimestamp()));
+        }
+        waitForCoordinatorToProcessActions(sourceCoordinator.getContext());
+    }
+
+    private long generateTimestamp() {
+        switch (timestampType) {
+            case MONOTONE_INCREASING:
+                return System.currentTimeMillis();
+            case RANDOMIZE_MILLISECONDS:
+                return System.currentTimeMillis() / 1000 * 1000 + RANDOM.nextInt(1000);
```

Comment:
Don't you actually benchmark here `RANDOM.nextInt()` and `System.currentTimeMillis()` instead of the `SourceCoordinator`? 🤔 `Random` is a very heavy class with lot's of overheads. `ThreadLocalRandom` is better, but there is no good replacement of `System.currentTimeMillis`, which is also quite heavy (syscall). 

Usually the solution is have some pre-generated small array of timestamps in the `setup` method, that's later used in the benchmark.